vm/devices/storage/disklayer_sqlite/src/lib.rs

```diff
@@ -0,0 +1,617 @@
+    ) -> Result<Self::Layer, Self::Error> {
+        let len = {
+            let lower_len = lower_layer_metadata
+                .as_ref()
+                .map(|m| m.sector_count * m.sector_size as u64);
+            self.format_dbhd
+                .len
+                .or(lower_len)
+                .context("no base layer to infer sector_count from")?
+        };
+        // FUTURE: make sector-size configurable
+        let sector_size = lower_layer_metadata.map(|x| x.sector_size).unwrap_or(512);
+
+        SqliteDiskLayer::new(
+            self.dbhd_path,
+            self.read_only,
+            Some(FormatParams {
+                logically_read_only: self.format_dbhd.logically_read_only,
+                len,
+                sector_size,
+            }),
+        )
+    }
+}
+
+impl LayerIo for SqliteDiskLayer {
+    fn layer_type(&self) -> &str {
+        "sqlite"
+    }
+
+    fn sector_count(&self) -> u64 {
+        self.meta.sector_count
+    }
+
+    fn sector_size(&self) -> u32 {
+        self.meta.sector_size
+    }
+
+    fn is_read_only(&self) -> bool {
+        self.read_only
+    }
+
+    fn disk_id(&self) -> Option<[u8; 16]> {
+        None
+    }
+
+    fn physical_sector_size(&self) -> u32 {
+        self.meta.sector_size
+    }
+
+    fn is_fua_respected(&self) -> bool {
+        true
+    }
+
+    async fn read(
+        &self,
+        buffers: &RequestBuffers<'_>,
+        sector: u64,
+        mut marker: SectorMarker<'_>,
+    ) -> Result<(), DiskError> {
+        let sector_count = (buffers.len() / self.meta.sector_size as usize) as u64;
+        let end_sector = sector + sector_count;
+        tracing::trace!(sector, sector_count, "read");
+        if end_sector > self.meta.sector_count {
+            return Err(DiskError::IllegalBlock);
+        }
+
+        let valid_sectors = unblock({
+            let conn = self.conn.clone();
+            let end_sector = sector + (buffers.len() as u64 / self.meta.sector_size as u64);
+            let sector_size = self.meta.sector_size;
+            move || read_sectors(conn, sector_size, sector, end_sector)
+        })
+        .await
+        .map_err(|e| DiskError::Io(std::io::Error::new(std::io::ErrorKind::Other, e)))?;
+
+        for (s, data) in valid_sectors {
+            let offset = (s - sector) as usize * self.meta.sector_size as usize;
+            let subrange = buffers.subrange(offset, self.meta.sector_size as usize);
+            let mut writer = subrange.writer();
+            match data {
+                SectorKind::AllZero => writer.zero(self.meta.sector_size as usize)?,
+                SectorKind::AllOne => writer.fill(0xff, self.meta.sector_size as usize)?,
+                SectorKind::Data(data) => writer.write(&data)?,
+            };
+
+            marker.set(s);
+        }
+
+        Ok(())
+    }
+
+    async fn write(
+        &self,
+        buffers: &RequestBuffers<'_>,
+        sector: u64,
+        _fua: bool,
```

Comment:
In WAL mode, writes can be lost due to power failure. `FUA` means this write must be flushed to disk. So, you either need to support this behavior somehow, or you need to change `is_fua_supported` above to `false`. I recommend the latter, because your FUA emulation will likely have to flush the disk anyway, and FUA is really only interesting to the guest if it causes a write-through write of the real disk cache and not a full flush of the disk cache.